### PR TITLE
WIP: Connection to last_letter once again successful

### PR DESCRIPTION
**Pull request by Georacer**
Communication between ardupilot and the last_letter simulator was broken for a few months now.
This is the first commit which brings things up to date.

### CODE REVIEW

**Comment by Georacer** [comment]
Hello everyone!

I started porting [last_letter](https://github.com/Georacer/last_letter) into Gazebo a couple of months back and I'd like to be able to work with master seamlessly.
I followed the Gazebo SITL template to make the modifications (which I suspect originally used the last_letter SITL template).

There are a couple of things that I'd like to ask, so this is mostly a WIP, although the above code compiles and runs.

Inline comments to follow.

File: libraries/SITL/SIM_last_letter.cpp [hunk 1]
```diff
@@ -53,39 +53,24 @@ last_letter::last_letter(const char *home_str, const char *_frame_str) :
...
-    if (child_pid != 0) {
-        return;
+    if (child_pid == 0) {
+      // in child
+      close(0);
+      open("/dev/null", O_RDONLY);
+      for (uint8_t i=3; i<100; i++) {
+          close(i);
+      }
+      int ret = execlp("roslaunch",
```
Comment (Georacer) [inline libraries/SITL/SIM_last_letter.cpp]: Followed the Gazebo SITL code to re-form the calling argument to comply with the new Gazebo simulator calling line of last_letter. This is expected to be expanded with more functionality, such as the HOME argument in the future.

File: libraries/SITL/SIM_last_letter.cpp [hunk 2]
```diff
@@ -146,6 +129,10 @@ void last_letter::update(const struct sitl_input &input)
     send_servos(input);
     recv_fdm(input);
     sync_frame_time();
+
+    update_position();
```
Comment (Georacer) [inline libraries/SITL/SIM_last_letter.cpp]: Originally `update_position` wasn't called. I added this after the Gazebo SITL example, but I'm not sure what it does. I suspect that it updates the position variables on the SITL backend where the APM HAL sensors read from.

File: libraries/SITL/SIM_last_letter.h [hunk 3]
```diff
@@ -41,7 +41,7 @@ class last_letter : public Aircraft {
     }
 
 private:
-    static const uint16_t fdm_port = 9002;
+    static const uint16_t fdm_port = 5002;
```
Comment (Georacer) [inline libraries/SITL/SIM_last_letter.h]: This is a big question mark for me.
Originally I used the 9002/3 ports for SITL-simulator communication.
Afterwards, I noticed that Gazebo-SITL uses 5002.
On the [SITL architecture page](http://ardupilot.org/dev/docs/sitl-simulator-software-in-the-loop.html) I see that 5502/3 is the proposed pair.

Is there a preferred/proposed pair of ports? Is there a developer recommendation?

**Comment by khancyr** [comment]
hi ! Maybe you are using an outdated gazebo pluggin. 
The current on is https://github.com/ArduPilot/ardupilot/blob/master/libraries/SITL/SIM_Gazebo.cpp . It works direclty with gazebo without using ROS as there is a plugin directly in gazebo (not yet in master)
Docs to install gazebo is here https://github.com/ArduPilot/ardupilot_wiki/pull/572 . I see you are using bullet engine ... So I will need to update the doc to include it to the build ^^ .

About ports: it is still 9002/3 for now. I hope to make it variable to enable multiple drone ! See PR about swarm on gazebo

**Comment by Georacer** [comment]
One more thing: currently the communication works (fdm packets) but the APM AHRS does not respond as it should and usually goes bad.

I think I send the correct accelerometer data; this will be my next goal to investigate.

**Comment by Georacer** [comment]
@khancyr re Bullet: I had a frustrating time with Gazebo-ODE. Any craft I would make would oscillate in acceleration +-1G, making any IMU recording unusable. This was mainly because of the solver steps for ground reactions and internal multibody foroces.
Bullet is extremely stable, but is lacking in API. As an example, it does not support `addRelativeForce()`, contrary to what it seems.

File: libraries/SITL/SIM_last_letter.h (hunk 3 above)
Comment (tridge) [inline libraries/SITL/SIM_last_letter.h]: this is completely up to you, as long as its documented I think either is fine

**Comment by tridge** [comment]
@Georacer to check sensors, try with AHRS_EKF_TYPE=10, that bypasses the EKF and you will be able to fly. If that works and it doesn't work with AHRS_EKF_TYPE=2 then it means the sensor data is not kinematically consistent or the timing is off

**Comment by tridge** [comment]
I'm going to merge this now as it fixes it to be usable again at least, then we can fix the remaining things in future PRs

**Pull request merged**